### ISSUE TITLE
Add Canine Model Config to AutoModelForCausalLM

### ISSUE THREAD

**Issue by khadiravana-belagavi** [huggingface/transformers#22066]
### Feature request

Kindly add a class such as https://github.com/huggingface/transformers/blob/a9bd5df16a46356463f2712dd8f6c109fa83d6f9/src/transformers/models/bert/modeling_bert.py#L1161

for the [Canine Model](https://huggingface.co/docs/transformers/model_doc/canine).

Basically, in the list of models available for CausalLM provided [here](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForCausalLM), the [canine](https://huggingface.co/docs/transformers/model_doc/canine) model isn't listed. Kindly add it. 

### Motivation

Currently unable to experiment with CanineConfig LM decoder using [this](https://huggingface.co/docs/transformers/model_doc/vision-encoder-decoder#transformers.VisionEncoderDecoderConfig.example) api. 

Snippet of code used:
 ```                                                                                                                                                                                                    
from transformers import ViTConfig, VisionEncoderDecoderConfig, VisionEncoderDecoderModel, CanineConfig
 # taken from https://huggingface.co/docs/transformers/v4.26.1/en/model_doc/vision-encoder-decoder#transformers.VisionEncoderDecoderConfig.example
config_encoder = ViTConfig()
config_decoder = CanineConfig()
config = VisionEncoderDecoderConfig.from_encoder_decoder_configs(config_encoder, config_decoder)
model = VisionEncoderDecoderModel(config=config)
```

### Your contribution

Not yet, currently.

**Comment by NielsRogge** [comment]
Hi,

CANINE doesn't support causal attention. It can only be used as an encoder.

**Comment by khadiravana-belagavi** [comment]
Thanks @NielsRogge  for pointing that out. Is there then, any language model similar to that of canine that processes the tokens at unicode character level. i.e. the tokenizer basically does 
```
uids = [ord(c) for c in string]
```

**Comment by NielsRogge** [comment]
You can leverage the decoder of [ByT5](https://huggingface.co/docs/transformers/model_doc/byt5), which is a byte-based model.

**Comment by khadiravana-belagavi** [comment]
Cool, that works. Thanks.

**Comment by khadiravana-belagavi** [comment]
@NielsRogge I think ByT5 while it does have the tokenization the way I wanted, it still cannot be used by the VisualEncoderDecoder API of hugging face - using the snippet like shown below:
```
from transformers import ViTConfig, VisionEncoderDecoderConfig, VisionEncoderDecoderModel, ByT5Config # this does not exist
# taken from https://huggingface.co/docs/transformers/v4.26.1/en/model_doc/vision-encoder-decoder#transformers.VisionEncoderDecoderConfig.example
config_encoder = ViTConfig()
config_decoder = ByT5Config() # this is what is desired.
config = VisionEncoderDecoderConfig.from_encoder_decoder_configs(config_encoder, config_decoder)
model = **VisionEncoderDecoderModel(config=config)**
```

**Comment by NielsRogge** [comment]
Hi @khadiravana-belagavi this is because T5/ByT5 is an encoder-decoder model. You would only need the decoder to combine it with a vision encoder. The vision encoder-decoder framework doesn't work out-of-the-box with T5/ByT5 at the moment as this would require us to define a new class that includes only the decoder + a language modeling head on top. 

Hence I'd recommend defining this class yourself and then provide it as decoder argument when instantiating a `VisionEncoderDecoderModel` class. The class could roughly look like this:
``` 
from transformers.models.t5.modeling_t5 import T5PreTrainedModel, T5Stack

class T5DecoderOnlyForCausalLM(T5PreTrainedModel):

  def __init__(self, config):
      self.decoder = T5Stack(config, self.shared)
      self.lm_head = nn.Linear(config.d_model, config.vocab_size, bias=False)
```
Then you can instantiate the model as follows:
```
from transformers import VisionEncoderDecoderModel, ViTModel

encoder = ViTModel.from_pretrained("google/vit-base-patch16-224")
decoder = T5DecoderOnlyForCausalLM.from_pretrained("t5-base")

model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
```

**Comment by khadiravana-belagavi** [comment]
Thanks @NielsRogge for the detailed response. However, I think this issue is still relevant, as although Canine is not a CausalLM, Bert is not as well. And the class `BertLMHeadModel` adds the necessary components for finetuning on CausalLM. Or is there anything specific to Canine - because canine is also a pretrained on a similar MLM task.

**Comment by NielsRogge** [comment]
@khadiravana-belagavi BERT can be adapted to be used as decoder (by simply using a causal attention mask rather than a bidirectional one). CANINE on the other hand cannot simply be adapted to work as decoder since it uses a different architecture composed of 3 Transformers.

**Comment by Biyani404198** [comment]
Hi @NielsRogge, I have been intending to use ByT5 as decoder too and I am getting the same error. Thanks for providing with the method to so.
`from transformers.models.t5.modeling_t5 import T5PreTrainedModel, T5Stack

class T5DecoderOnlyForCausalLM(T5PreTrainedModel):

  def __init__(self, config):
      self.shared = nn.Embedding(config.vocab_size, config.d_model)
      self.decoder = T5Stack(config, self.shared)
      self.lm_head = nn.Linear(config.d_model, config.vocab_size, bias=False)`

`from transformers import VisionEncoderDecoderModel, ViTModel

encoder = ViTModel.from_pretrained("google/vit-base-patch16-224")
decoder = T5DecoderOnlyForCausalLM.from_pretrained("t5-base")

model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)`

Can you provide with the detailed code or any reference so that I can accurately create the complete T5DecoderOnlyForCausalLM class and the weights of the decoder are properly instantiated.
